# Let's start with Python programming

In [1]:
x = 7 # Command/Program

# Shift + Enter = Execute cell

In [2]:
var = 707

In [3]:
var # Always use lowercase to avoid time consuming errors

707

In [4]:
x,y = (1,5) # use parentheses to improve readability of code

In [5]:
x

1

In [6]:
y

5

In [7]:
# There can be different types in python
# There can be integers (finite range)

x = 11
y = -7


In [8]:
#Float 
# Add a . or decimal


In [9]:
# Strings in single or double quotes
"Hello World"

'Hello World'

In [10]:
#Boolean
# True or false
True

True

In [11]:
# To see what type is retured wrap type(obj)
type("Hello World")

str

In [12]:
# Integers can be either positive or negative

t = 7 # assigning 7 to variable t

In [13]:
type(t)

int

In [14]:
t2 = 7.907 # Any decimal python interperets as float

In [15]:
type(t2)

float

In [16]:
import sys

In [17]:
sys.float_info

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

In [18]:
int(t2) # Converts obj to integers rounded to the nears integer

7

In [19]:
float(t)

7.0

# Booleans

In [20]:
abc = True

In [21]:
type(abc)

bool

# Boolean

In [22]:
#Booleans can be True or False

In [23]:
abcd = True

In [24]:
abcd

True

In [25]:
# Python is case sensitive boolean must have an uppercase letter

# Strings

In [26]:
# How to type text in python
# You must specify with a ' ' or " "
'sam'

'sam'

In [27]:
type('sam')

str

In [28]:
# If you want to print the string without "" in text
print("Hey dude") # Use print statement to remove the quotes

Hey dude


In [29]:
'I love Trading'

'I love Trading'

In [30]:
my_name = 'Josh'

In [31]:
print(my_name)

Josh


# Operations

In [32]:
# When you have options 
# When working with strings add a space before or in your variable 
"My name is" + my_name



'My name isJosh'

In [33]:
"My name is " + my_name

'My name is Josh'

In [34]:
money = 1000000000000.99

In [35]:
my_name + " has " + str(money) 

'Josh has 1000000000000.99'

In [36]:
float(str(money))

1000000000000.99

# Numpy

Reasons to use numpy
* Massive code machine learning, deep learning models runtime can be high a long loops
* Uses less memory (Efficency)
* More convenient (Math operations) Multiplying matricies
* Funconality (Can be used with pandas and many API's)

In [37]:
import numpy as np

In [38]:
list_a = [5,6,7,8,9,10]
type(list_a)

list

In [39]:
np.array(list_a)

array([ 5,  6,  7,  8,  9, 10])

In [40]:
arr = np.array(list_a) # Convert to a list and stored as a Numpy array

In [41]:
type(arr)  # You must use numpy in your daily life

numpy.ndarray

# Function arange

In [42]:
# You can create an array just by defining your steps
array = np.arange(6,10) # Three arguments start, stop, step
# Stop is always excludede
array

array([6, 7, 8, 9])

In [43]:
array = np.arange(0,10,2)

In [44]:
array

array([0, 2, 4, 6, 8])

In [45]:
# If you want 10 to be included increase the stop by 1
array = np.arange(0,11,2)
array

array([ 0,  2,  4,  6,  8, 10])

# Np Zero, One

In [46]:
size = (7,3) # Rows * cols

In [47]:
matrix = np.ones(size) # Created matrix with 7 rows and 3 columns

In [48]:
print(matrix)

[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]


In [49]:
matrix = np.zeros((4,2))

In [50]:
print(matrix)

[[0. 0.]
 [0. 0.]
 [0. 0.]
 [0. 0.]]


# Random matrix

In [51]:
random_matrix = np.random.randint(0,50)

In [52]:
random_matrix

19

In [53]:
random_matrix = np.random.randint(0,50,size)

In [54]:
random_matrix

array([[17, 30,  4],
       [ 5, 31, 27],
       [29, 15, 48],
       [11,  3, 11],
       [40,  0, 12],
       [41, 10, 41],
       [20,  3,  1]], dtype=int32)

# Line space
* ax + b
* Ask numpy to provide values good for testing values or having values with the same space
* You need 3 paranetersmm A, B, Nb

In [55]:
# Create linearly spaced array
x = 0
y= 10
nb_val = 5
matrix = np.linspace(x,y,nb_val)
print(matrix)

[ 0.   2.5  5.   7.5 10. ]


# Operations

In [56]:
arr = np.random.randint(0,1000,80)

In [57]:
print(arr)

[787 612  28  23 241 628 546 437 827 362 752  34 757 800 697 304 446 542
 606  71 481 860 601 898 461 570 399 481 225 608 461 450 479 338 165 743
 382 532 624 687 608 359 419 595 667 477 161 608  54 536 898 931 509 458
 178 862  29 412 208 581 281 945 948 983 656 316 934 565  25 920 437  76
 638 618 346 661 288 939  79 340]


In [58]:
# Simple operations

print(arr.min()) # prints minimum value in the array
print(arr.max()) # prints maximum value in the array
print(arr.mean()) # prints mean value in the array (avg of the list)
print(arr.std()) # prints the std value in the array
print(arr.var()) # prints the variance in the array
print(arr.std()**2) # Prints the std deviation which is the sqrt(var)

23
983
506.125
261.8653458841013
68573.459375
68573.45937500002


In [59]:
print(f"The variance of the array is {arr.var()}")

The variance of the array is 68573.459375


* 2D 3D array
* Filter and locate inside an array
* Seed function

In [60]:
arr = np.random.randint(0,1000,24)
print(arr)

[355 890 510 973 750 225 351 949 682  21 213 875 438 904 499 722 206  48
 400 667  69 512 848 964]


In [82]:
mat = arr.reshape(6,4)# Number of rows and # of cols
print(mat)

[[355 890 510 973]
 [750 225 351 949]
 [682  21 213 875]
 [438 904 499 722]
 [206  48 400 667]
 [ 69 512 848 964]]


In an array or matrix begins at row 0 and col 0

*array([[355, 890, 510, 973],0
       [750, 225, 351, 949],1
       [682,  21, 213, 875],2
       [438, 904, 499, 722],3
       [206,  48, 400, 667],4
       [ 69, 512, 848, 964]],5 dtype=int32)

In [85]:
# 3D

matix = arr.reshape(3,2,4) # This must = 24
print(matix)

[[[355 890 510 973]
  [750 225 351 949]]

 [[682  21 213 875]
  [438 904 499 722]]

 [[206  48 400 667]
  [ 69 512 848 964]]]


In [89]:
# 4D Array

matix = arr.reshape(3,2,2,2) # This must = 24
print(matix)

[[[[355 890]
   [510 973]]

  [[750 225]
   [351 949]]]


 [[[682  21]
   [213 875]]

  [[438 904]
   [499 722]]]


 [[[206  48]
   [400 667]]

  [[ 69 512]
   [848 964]]]]


# locate

In [88]:
matix[2,1,1,1]

np.int32(964)

In [68]:
mat[2,3] #shape -1 otherwise it is out of bounds
print(mat[2,3])

875


In [69]:
# Filtering will convert matrix into boolean

In [70]:
mat > 100

array([[ True,  True,  True,  True],
       [ True,  True,  True,  True],
       [ True, False,  True,  True],
       [ True,  True,  True,  True],
       [ True, False,  True,  True],
       [False,  True,  True,  True]])

In [74]:
# To filter values in an array true/false matrix use a array[array > value]
mat[mat > 500] # Returns an array of the items in the array with a value > 500 

array([890, 510, 973, 750, 949, 682, 875, 904, 722, 667, 512, 848, 964],
      dtype=int32)

# Seed function

In [75]:
np.random.seed(707)

In [ ]:
random_matrix = np.random.randint(0,100,10) # Seed creates consistiny in the random results
random_matrix

array([67,  2, 86, 91, 90, 74, 73, 20, 45, 28], dtype=int32)



---

# 📘 **NumPy Multidimensional Arrays: 1D → 5D Complete Cheat Sheet**

---

# 🟦 **1. Dimensional Overview (1D → 5D)**

| Dimension | Shape Example | Meaning | Mental Model |
|----------|----------------|---------|---------------|
| **1D** | `(n,)` | A list of numbers | Line |
| **2D** | `(rows, cols)` | Matrix / table | Spreadsheet |
| **3D** | `(blocks, rows, cols)` | Stack of matrices | Book of pages |
| **4D** | `(groups, blocks, rows, cols)` | Stack of 3D arrays | Bookshelf of books |
| **5D** | `(sets, groups, blocks, rows, cols)` | Stack of 4D arrays | Library of shelves |

---

# 🟩 **2. Indexing Cheat Sheet (1D → 5D)**

## **1D Indexing**
```python
arr[i]
```

## **2D Indexing**
```python
arr[row, col]
arr[row][col]
```

## **3D Indexing**
Shape: `(B, R, C)`
```python
arr[block, row, col]
```

Examples:
```python
arr[0]        # block 0 → shape (R, C)
arr[0, 1]     # row 1 of block 0 → shape (C,)
arr[0, 1, 2]  # element
```

## **4D Indexing**
Shape: `(G, B, R, C)`
```python
arr[group, block, row, col]
```

Examples:
```python
arr[1]            # group 1 → shape (B, R, C)
arr[1, 0]         # block 0 of group 1 → shape (R, C)
arr[1, 0, 1]      # row 1 → shape (C,)
arr[1, 0, 1, 3]   # element
```

## **5D Indexing**
Shape: `(S, G, B, R, C)`
```python
arr[set, group, block, row, col]
```

Examples:
```python
arr[0]               # set 0 → shape (G, B, R, C)
arr[0, 1]            # group 1 → shape (B, R, C)
arr[0, 1, 2]         # block 2 → shape (R, C)
arr[0, 1, 2, 0]      # row 0 → shape (C,)
arr[0, 1, 2, 0, 3]   # element
```

---

# 🟧 **3. Visual Diagram: How a 4D Array Is Structured**

Let’s visualize a 4D array with shape `(2, 3, 2, 4)`:

```
Group 0
 ├── Block 0
 │    ├── Row 0: [....]
 │    └── Row 1: [....]
 ├── Block 1
 │    ├── Row 0: [....]
 │    └── Row 1: [....]
 └── Block 2
      ├── Row 0: [....]
      └── Row 1: [....]

Group 1
 ├── Block 0
 │    ├── Row 0: [....]
 │    └── Row 1: [....]
 ├── Block 1
 │    ├── Row 0: [....]
 │    └── Row 1: [....]
 └── Block 2
      ├── Row 0: [....]
      └── Row 1: [....]
```

**Interpretation:**  
A 4D array is a **collection of groups**, each containing **blocks**, each containing **rows**, each containing **columns**.

---

# 🟨 **4. NumPy → PyTorch Tensor Mapping**

| Concept | NumPy | PyTorch |
|--------|--------|----------|
| Create array/tensor | `np.array([...])` | `torch.tensor([...])` |
| Shape | `arr.shape` | `tensor.shape` |
| Reshape | `arr.reshape(...)` | `tensor.reshape(...)` |
| Add dimension | `np.expand_dims(arr, axis)` | `tensor.unsqueeze(axis)` |
| Remove dimension | `np.squeeze(arr)` | `tensor.squeeze()` |
| Transpose | `arr.transpose(...)` | `tensor.permute(...)` |
| 4D image batch | `(N, H, W, C)` | `(N, C, H, W)` |
| 5D video batch | `(N, T, H, W, C)` | `(N, C, T, H, W)` |

**Key difference:**  
PyTorch uses **channels‑first**, NumPy uses **channels‑last** by default.

---

# 🟪 **5. Reshape Rules (Critical)**

### ✔ Total number of elements must stay the same
```python
arr.size == np.prod(new_shape)
```

### ✔ NumPy fills row‑major (C‑order)
Meaning: last dimension changes fastest.

### ✔ Use `-1` to auto‑infer
```python
arr.reshape(3, -1)  # NumPy figures out the missing dimension
```

---

# 🟫 **6. Slicing Patterns (All Dimensions)**

### **General rule**
```
arr[a:b, c:d, e:f, ...]
```

### Examples

#### 3D
```python
arr[:, 0]      # all blocks, row 0
arr[1, :, :]   # block 1, all rows/cols
```

#### 4D
```python
arr[:, 0, :, :]   # all groups, block 0
arr[1, :, :, 2]   # group 1, all blocks/rows, column 2
```

#### 5D
```python
arr[:, :, 0]      # all sets, all groups, block 0
```

---

# 🟦 **7. Practice Exercises (with Solutions)**

---

## **Exercise 1 — Create a 3D array**
Create a 3D array of shape `(3, 2, 4)` from `np.arange(24)`.

**Solution**
```python
arr = np.arange(24).reshape(3, 2, 4)
```

---

## **Exercise 2 — Indexing**
Given shape `(3, 2, 4)`, extract:

1. Block 2  
2. Row 1 of block 0  
3. Element at block 1, row 0, col 3  

**Solution**
```python
arr[2]
arr[0, 1]
arr[1, 0, 3]
```

---

## **Exercise 3 — 4D indexing**
Given shape `(2, 3, 2, 4)`:

Extract:

1. Group 1  
2. Block 2 of group 0  
3. Row 1, col 2 of group 1, block 0  

**Solution**
```python
arr[1]
arr[0, 2]
arr[1, 0, 1, 2]
```

---

## **Exercise 4 — Reshape to 5D**
Turn `np.arange(120)` into shape `(2, 3, 2, 2, 5)`.

**Solution**
```python
arr = np.arange(120).reshape(2, 3, 2, 2, 5)
```

---

## **Exercise 5 — NumPy → PyTorch**
Convert a NumPy image batch `(N, H, W, C)` to PyTorch `(N, C, H, W)`.

**Solution**
```python
torch_tensor = torch.from_numpy(arr).permute(0, 3, 1, 2)
```

---

